### DISCUSSION TITLE
Possibly moving department from Solomon to Boris: questions and issues

### DISCUSSION THREAD

**Discussion by ferdinandyb** [olivierfriard/BORIS#487] [Ideas]
A little background: I'm creating this issue as an initial question and feature request on behalf of our department, where we are considering to move from [Solomon Coder](https://solomon.andraspeter.com/) to BORIS. We are around ~40 staff and with our students around ~100 potential regular users (Eötvös Loránd University, Department of Ethology and related research groups), who have been using Solomon heavily up to now, but as it is being discontinued, we are looking for an alternative. Writing our own software also came to mind, but since we found this open source project, we believe our efforts would be better spent on BORIS, as it seems to be a great project.

Below I will have some questions about things that might already work, but we couldn't figure and some feature requests. I'm dumping everything here, since I'm unfamiliar with the development process, but later I'm very happy to break out everything into different issues and if you can help me with how the project is organised and some pointers about where best to implement things, I would be
happy to also contribute via PR-s.

Our issues/requests:

# Show coded behaviour in sync with video during coding

Visual feedback during coding would be very helpful, either somehow drawn over the video, or just a running table or something that scrolls in sync with the video.

# Ability do draw lines in videos

In Solomon you could draw lines over the video (and save them), which is a heavily used feature for helping to decide if the animal crossed an imaginary  boundary or not.

# More flexibility in presenting data from external files

The current way of plotting data from external files is rather limited. I'm not sure what is the exact limitation of having only two plots, but actually maybe having the ability to plot multiple columns from a single file onto a single plot might be more useful, than a third plotwindow.

# Sharing of .boris project files is a bit cumbersome

I really like the idea of having everything in a single json for a single project, but it is a bit unflexible, if multiple are working on multiple
observations since:

- the paths to videos and data files are hardcoded absolute paths (won't work on   somebody else's system)
- you end up with say 3 .boris files each having a subset of observations

Having some merging capabilities and possibility to add relative paths would solve this I think. 

# Latency calculations

The desired method would be that the user marks a point in time and the software calculates latencies from that point for all behaviours coded.

I hope most of these we just missed, and I reiterate, that with help I'm happy to make PR-s if you think these problems can or should be solved. Thanks!

**Comment by olivierfriard** [comment]
Hello,

Sorry for the delay. Did you start to use BORIS?

For sharing a project, you can decide to do not store the entire path of your media.
In this case your media must be in the same directory than your project file.
I am working on this to introduce a relative path to media (relative to the project file)
Best,
Olivier

**Comment by ferdinandyb** [comment]
No worries :) Yes we did. We also managed to figure out pathless media solution. I think our current main issue are the latency and a visual representation of what was coded on the video. Do you have a list of features you are planning to implement or a roadmap? Are you open to contributions? I'd be happy to assist in bringing features to life that we need/want if it fits your plans.

**Comment by olivierfriard** [comment]
Hi,
since v.8.1.2 (based on the MPV media player) there is the possibility to add media/data files saving the path relatively to the BORIS project file.
The code of v. 8.1.2 was also completely refactored to be more open to collaborations.

I am working to the latency. I plan to add the latencies after selecting an arbitrary point event (or the end of a state event).
Should it be OK for you?

**Comment by tamsa42** [comment]
Hi @olivierfriard, 

thanks a lot for working on this! 

in Solomon the logic of latency measurement is similar as you wrote, that you can set up any coded behaviour as marker (point like or state) as a 0 point then add a selection of other behaviours to measure, then it will calculate the time it takes till the occurrences. for the censored events (aka behaviour not happening within a timeframe or till the next occurrence of the marker) you can define a value and it will log that in the output. another handy feature is that you can select whether you want to save only the latency of the first occurrence of a behaviour within the timeframe, or all occurrences. 

Maybe an example helps to imagine. in a playback study we play different sounds to the subject and want to compare the latency of orientation between the different sound categories. in this case, we would use the start of the sound playbacks as the 0 point and want to measure the time it takes to the first timepoint of orientation till the start of the next playback. if there is no orientation the time between the two playback start points should be saved. 

would the latency measurement possible to be implemented like this or using a similar logic?

**Comment by olivierfriard** [comment]
Hi,
just implemented the latency function in the pre-release of v. 8.3 (See Analyze > Latency).
It is a first version of the function that will be useful to work on it.
Everything is not implemented yet.
Can you give me some advice for the output (now it is a text output but I plan to write a spreadsheet like output)?

You can download the Windows version at:
https://drive.google.com/file/d/1PnqhLu6ExvGuhI2wjig8NrKrUtqmK40q/view?usp=sharing
Tell me if you are using Linux, no MacOS version will be available.

**Reply by ferdinandyb** [reply]
Hi, thanks, we'll try to test it soon. Most of use windows, but anyway, we can get it from the repo directly, I guess it's this branch: https://github.com/olivierfriard/BORIS/tree/mpv
